staging/src/k8s.io/client-go/tools/watch/retrywatcher.go

```diff
@@ -0,0 +1,232 @@
+					})
+					// We have to abort here because this might cause lastResourceVersion inconsistency by skipping a potential RV with valid data!
+					return true, 0
+				}
+
+				resourceVersion := metaObject.GetResourceVersion()
+				if resourceVersion == "" {
+					_ = rw.send(watch.Event{
+						Type:   watch.Error,
+						Object: &apierrors.NewInternalError(fmt.Errorf("retryWatcher: object %#v doesn't support resourceVersion", event.Object)).ErrStatus,
+					})
+					// We have to abort here because this might cause lastResourceVersion inconsistency by skipping a potential RV with valid data!
+					return true, 0
+				}
+
+				// All is fine; send the event and update lastResourceVersion
+				ok = rw.send(event)
+				if !ok {
+					return true, 0
+				}
+				rw.lastResourceVersion = resourceVersion
+
+				continue
+
+			case watch.Error:
+				status, ok := event.Object.(*metav1.Status)
+				if !ok {
+					klog.Error(spew.Sprintf("RetryWatcher received an error which is not *metav1.Status but %#+v", event.Object))
+					// Retry unknown errors
+					return false, 0
+				}
+
+				switch status.Code {
+				case http.StatusGone:
+					// Never retry RV too old errors
+					_ = rw.send(event)
+					return true, 0
+
+				case http.StatusGatewayTimeout:
+					if status.Details == nil {
+						// Retry
+						return false, 0
+					}
+
+					// Retry after delay
+					return false, time.Duration(status.Details.RetryAfterSeconds) * time.Second
+
+				case http.StatusInternalServerError:
+					// Retry
+					return false, 0
+
+				default:
+					// We retry by default. RetryWatcher is meant to proceed unless it is certain
+					// that it can't. If we are not certain, we proceed with retry and leave it
+					// up to the user to timeout if needed.
+
+					// Log here so we have a record of hitting the unexpected error
+					// and we can whitelist some error codes if we missed any that are expected.
+					klog.V(5).Info(spew.Sprintf("RetryWatcher - retrying after unexpected error: %#+v", event.Object))
+
+					return false, 0
+				}
+
+			default:
+				klog.Errorf("RetryWatcher failed to recognize Event type %q", event.Type)
+				_ = rw.send(watch.Event{
+					Type:   watch.Error,
+					Object: &apierrors.NewInternalError(fmt.Errorf("retryWatcher failed to recognize Event type %q", event.Type)).ErrStatus,
+				})
+				// We are unable to restart the watch and have to stop the loop or this might cause lastResourceVersion inconsistency by skipping a potential RV with valid data!
+				return true, 0
+			}
+		}
+	}
+}
+
+// receive reads the result from a watcher, restarting it if necessary.
+func (rw *RetryWatcher) receive() {
+	defer close(rw.doneChan)
+
+	klog.V(4).Info("Starting RetryWatcher.")
+	retryAfter := 0 * time.Second
+	var done bool
+	for {
+		select {
+		case <-rw.stopChan:
+			klog.V(4).Info("Stopping RetryWatcher.")
+			return
+		case <-time.After(retryAfter):
+			done, retryAfter = rw.doReceive()
```

Comment:
other loops that re-establish watches do so at a 1 second interval. how are we avoiding hot-looping here?

Reply:
are those (1s) mostly controller loops? my thinking is - watch getting closed is a normal occurrence and we should reestablish it as fast as we can, as this should behave the same way as a singe watch and reestablishing it shouldn't be observed nor cause artificial delays, at least for the first time - maybe we need to limit consequent failures or we are good with leaving this on client rate-limiter?

Reply:
delays are now added in the inner function

Reply:
I still see several places it does `return false, 0`